Given two sorted arrays $nums1$ and $nums2$ of same size $n$ respectively, return the median of the two sorted arrays.

The overall run time complexity should be $O(log(n))$.

*Example*<br />

$nums1 = [1, 2]$<br />
$nums2 = [3, 4]$<br />
Merged array is $[1, 2, 3, 4]$ and median is $2.5$

*Code*

In [8]:
def find_median_sorted_arrays(nums1, nums2):
    if not nums1 and not nums2:
        return 0

    l1 = len(nums1)
    l2 = len(nums2)

    if l1 == 1 and l1 == l2:
        return (nums1[0] + nums2[0]) / 2

    if l1 == 2 and l2 == 2:
        return (max(nums1[0], nums2[0]) + min(nums1[1], nums2[1])) / 2

    get_mean = lambda arr: (arr[len(arr)//2] + arr[-(len(arr)//2 + 1)]) / 2

    m1 = get_mean(nums1)
    m2 = get_mean(nums2)

    if m1 == m2:
        return m1

    if m1 > m2:
        sub1 = nums1[:l1//2 + 1]
        sub2 = nums2[l2//2 - (1 if l2 % 2 == 0 else 0):]
    else:
        sub1 = nums1[l1//2 - (1 if l1 % 2 == 0 else 0):]
        sub2 = nums2[:l2//2 + 1]

    return find_median_sorted_arrays(sub1, sub2)

In [10]:
m = find_median_sorted_arrays([1, 12, 15, 26, 38, 40], [2, 13, 17, 30, 40, 45])
print(m)

21.5


*Explanation*

The method works by first getting medians of the two sorted arrays and then comparing them.

1. Calculate the medians $m1$ and $m2$ of the input arrays $nums1$ and $nums2$ respectively.
2. If $m1$ and $m2$ both are equal then we are done. So, return $m1$ (or $m2$).
3. If $m1$ is greater than $m2$, then median is present in one of the below two subarrays:<br />
    a) from first element of $nums1$ to $m1$ ($nums1[0 \dots n/2]$);<br />
    b) from $m2$ to last element of $nums2$ ($nums2[n/2 \dots n-1]$).<br />
4. If $m2$ is greater than $m1$, then median is present in one of the below two subarrays:<br />
    a) from $m1$ to last element of $nums1$ ($nums1[n/2 \dots n-1]$);<br />
    b) from first element of $nums2$ to $m2$ ($nums2[0 \dots n/2]$).<br />
5. Repeat the above process until size of both the subarrays becomes $2$.
6. If size of the two arrays is $2$ then use below formula to get the median:<br />
    $m = (max(nums1[0], nums2[0]) + min(nums1[1], nums2[1])) / 2$
    
*Example*

$nums1 = {1, 12, 15, 26, 38, 40} \\
nums2 = {2, 13, 17, 30, 40, 45}$

For above two arrays $m1 = 20.5$ and $m2 = 23.5$. $m2$ is greater than $m1$. So, median is present in one of the following two subarrays:
$nums1 = {15, 26, 38, 40} \\
nums2 = {2, 13, 17, 30}$

For above two subarrays $m1 = 32$ and $m2 = 15$. Now $m1$ is greater than $m2$. So:
$nums1 = {15, 26, 38} \\
nums2 = {13, 17, 30}$

For updated subarrays $m1 = 26$ and $m2 = 17$. $m1$ is greater than $m2$:
$nums1 = {15, 26} \\
nums2 = {17, 30}$

Now size is $2$, so $m = (max(15, 17) + min(26, 30)) / 2 = (17 + 26) / 2 = 21.5$.

Get median in a regular way:<br />
${1, 2, 12, 13, 15, \textbf{17}, \textbf{26}, 30, 38, 40, 40, 45}$<br />
$m = (17 + 26) / 2 = 21.5$.